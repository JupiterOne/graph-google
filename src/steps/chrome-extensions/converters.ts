import generateEntityKey from '../../utils/generateEntityKey';
import { entities } from '../../constants';
import {
  Entity,
  RelationshipClass,
  createDirectRelationship,
  createIntegrationEntity,
} from '@jupiterone/integration-sdk-core';
import { chromemanagement_v1 } from 'googleapis';

export function createChromeExtensionEntity(
  data: chromemanagement_v1.Schema$GoogleChromeManagementV1InstalledApp,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: generateEntityKey(entities.CHROME_EXTENSION._type, data.appId),
        _type: entities.CHROME_EXTENSION._type,
        _class: entities.CHROME_EXTENSION._class,
        id: data.appId || 'UNKNOWN',
        displayName: data.displayName || data.appId || 'UNKNOWN',
        name: data.displayName,
        description: data.description,
        installType: data.appInstallType,
        source: data.appSource,
        type: data.appType,
        deviceCount: data.browserDeviceCount ?? null,
        disabled: data.disabled,
        uri: data.homepageUri,
        permissions: data.permissions?.join(',') || '',
      },
    },
  });
}

export function createChromeOsDeviceInstalledChromeExtensionRelationship(params: {
  deviceEntity: Entity;
  extensionEntity: Entity;
}) {
  return createDirectRelationship({
    _class: RelationshipClass.INSTALLED,
    from: params.deviceEntity,
    to: params.extensionEntity,
  });
}
