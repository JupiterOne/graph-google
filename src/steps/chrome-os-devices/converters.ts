import generateEntityKey from '../../utils/generateEntityKey';
import { admin_directory_v1 } from 'googleapis';
import { entities } from '../../constants';
import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

export function createChromeOSDeviceEntity(
  data: admin_directory_v1.Schema$ChromeOsDevice,
) {
  const deviceName = data.deviceId as string;

  const serialNumber = data.serialNumber;
  const lastSeenOn = parseTimePropertyValue(data.lastSync);

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: generateEntityKey(entities.CHROME_OS_DEVICE._type, deviceName),
        _type: entities.CHROME_OS_DEVICE._type,
        _class: entities.CHROME_OS_DEVICE._class,
        id: deviceName,
        category: 'chrome-os-device',
        make: data.model,
        name: deviceName,
        status: data.status,
        model: data.model,
        serial: data.serialNumber,
        serialNumber,
        version: data.osVersion,
        supportEndDate: data.supportEndDate,
        platformVersion: data.platformVersion,
        firmwareVersion: data.firmwareVersion,
        macAddress: data.macAddress,
        lastSync: lastSeenOn,
        lastSeenOn,
        lastEnrollementTime: parseTimePropertyValue(data.lastEnrollmentTime),
        bootMode: data.bootMode,
        tpmFamily: data.tpmVersionInfo?.family,
        tpmFirmwareVersion: data.tpmVersionInfo?.firmwareVersion,
        tpmManufacturer: data.tpmVersionInfo?.manufacturer,
        tpmSpecLevel: data.tpmVersionInfo?.specLevel,
        tpmModel: data.tpmVersionInfo?.tpmModel,
        tpmVendorId: data.tpmVersionInfo?.vendorSpecific,
      },
    },
  });
}

export function createAccountManagesChromeOSDeviceRelationship(params: {
  accountEntity: Entity;
  deviceEntity: Entity;
}) {
  return createDirectRelationship({
    _class: RelationshipClass.MANAGES,
    from: params.accountEntity,
    to: params.deviceEntity,
  });
}
