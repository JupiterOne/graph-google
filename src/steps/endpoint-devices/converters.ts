import generateEntityKey from '../../utils/generateEntityKey';
import { cloudidentity_v1 } from 'googleapis';
import { entities, relationships } from '../../constants';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
  Entity,
  createDirectRelationship,
} from '@jupiterone/integration-sdk-core';

export function createDeviceEntity(
  data: cloudidentity_v1.Schema$GoogleAppsCloudidentityDevicesV1Device,
) {
  const serialNumber = data.serialNumber || '';
  const lastSeenOn = parseTimePropertyValue(data.lastSyncTime);

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: generateEntityKey(entities.DEVICE._type, data.name),
        _type: entities.DEVICE._type,
        _class: entities.DEVICE._class,
        id: data.name!,
        name: data.name,
        displayName: data.name!,
        createdOn: parseTimePropertyValue(data.createTime),
        lastSyncedOn: lastSeenOn,
        ownerType: data.ownerType,
        model: data.model || '',
        osVersion: data.osVersion,
        deviceType: data.deviceType,
        securityPatchedOn: parseTimePropertyValue(data.securityPatchTime),
        encryptionState: data.encryptionState,
        encrypted: data.encryptionState === 'ENCRYPTED',
        category: 'endpoint',
        make: data.manufacturer || '',
        serial: serialNumber,
        serialNumber,
        // Monkeypatch: deviceId field is not included in type but is included in response.
        // ** deviceId is a required field in J1 Device schema.
        deviceId: (data as any).deviceId || '',
        byod: data.ownerType === 'BYOD',
        assetTag: data.assetTag,
        imei: data.imei,
        meid: data.meid,
        wifiMacAddresses: data.wifiMacAddresses,
        macAddress: data.wifiMacAddresses,
        networkOperator: data.networkOperator,
        releaseVersion: data.releaseVersion,
        brand: data.brand,
        buildNumber: data.buildNumber,
        kernelVersion: data.kernelVersion,
        basebandVersion: data.basebandVersion,
        enabledDeveloperOptions: data.enabledDeveloperOptions,
        otherAccounts: data.otherAccounts,
        enabledUsbDebugging: data.enabledUsbDebugging,
        bootloaderVersion: data.bootloaderVersion,
        managementState: data.managementState,
        compromisedState: data.compromisedState,
        androidEnabledUnknownSources:
          data.androidSpecificAttributes?.enabledUnknownSources,
        androidSupportsWorkProfile:
          data.androidSpecificAttributes?.supportsWorkProfile,
        androidOwnerProfileAccount:
          data.androidSpecificAttributes?.ownerProfileAccount,
        androidOwnershipPrivilege:
          data.androidSpecificAttributes?.ownershipPrivilege,
      },
    },
  });
}

export function createAccountManagesDeviceRelationship(params: {
  accountEntity: Entity;
  deviceEntity: Entity;
}) {
  return createDirectRelationship({
    _class: relationships.ACCOUNT_MANAGES_DEVICE._class,
    from: params.accountEntity,
    to: params.deviceEntity,
  });
}
