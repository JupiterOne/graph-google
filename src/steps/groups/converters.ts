import generateEntityKey from '../../utils/generateEntityKey';
import {
  createIntegrationEntity,
  createDirectRelationship,
  RelationshipClass,
  Entity,
  createMappedRelationship,
  RelationshipDirection,
} from '@jupiterone/integration-sdk-core';
import { admin_directory_v1 } from 'googleapis';
import { entities, relationships } from '../../constants';

export enum MemberType {
  CUSTOMER = 'CUSTOMER',
  EXTERNAL = 'EXTERNAL',
  GROUP = 'GROUP',
  USER = 'USER',
}

export function createGroupEntity(data: admin_directory_v1.Schema$Group) {
  const groupId = data.id as string;

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: generateEntityKey(entities.GROUP._type, groupId),
        _type: entities.GROUP._type,
        _class: entities.GROUP._class,
        id: groupId,
        adminCreated: data.adminCreated,
        directMembersCount: data.directMembersCount,
        displayName: data.name as string,
        email: data.email,
        kind: data.kind,
        name: data.name,
        description: data.description,
      },
    },
  });
}

export function createGroupHasGroupMappedRelationship(
  sourceGroupEntity: Entity,
  groupMember: admin_directory_v1.Schema$Member,
) {
  const targetGroupEntityKey = generateEntityKey(
    'group',
    groupMember.id as string, // Google doesn't ensure the uniqueness of emails in their groups.
  );

  return createMappedRelationship({
    _class: RelationshipClass.HAS,
    _key: `${sourceGroupEntity._key}_has_${targetGroupEntityKey}`,
    _type: relationships.GROUP_HAS_USER._type,
    _mapping: {
      relationshipDirection: RelationshipDirection.FORWARD,
      sourceEntityKey: sourceGroupEntity._key,
      skipTargetCreation: false,
      targetFilterKeys: [['_type', 'email']],
      targetEntity: {
        _type: entities.GROUP._type,
        email: groupMember.email,
      },
    },
  });
}

export function createGroupHasGroupRelationship(params: {
  sourceGroupEntity: Entity;
  targetGroupEntity: Entity;
  groupMember: admin_directory_v1.Schema$Member;
}) {
  const { sourceGroupEntity, targetGroupEntity, groupMember } = params;

  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: sourceGroupEntity,
    to: targetGroupEntity,
    properties: {
      deliverySettings: groupMember.delivery_settings,
      email: groupMember.email,
      id: groupMember.id,
      kind: groupMember.kind,
      role: groupMember.role,
      status: groupMember.status,
      type: groupMember.type,
    },
  });
}

// This means that there is no user found in this integration with user groupMember.id -- ie: this is an EXTERNAL group member
export function createGroupHasUserMappedRelationship(
  sourceGroupEntity: Entity,
  groupMember: admin_directory_v1.Schema$Member,
) {
  const targetUserEntityKey = generateEntityKey(
    'user',
    groupMember.id as string, // Google doesn't ensure the uniqueness of emails in their groups.
  );

  // Create a mapped relationship
  return createMappedRelationship({
    _class: RelationshipClass.HAS,
    _key: `${sourceGroupEntity._key}_has_${targetUserEntityKey}`,
    _type: relationships.GROUP_HAS_USER._type,
    _mapping: {
      relationshipDirection: RelationshipDirection.FORWARD,
      sourceEntityKey: sourceGroupEntity._key,
      skipTargetCreation: false,
      targetFilterKeys: [['_type', 'email']],
      targetEntity: {
        _type: entities.USER._type,
        email: groupMember.email,
      },
    },
  });
}

export function createGroupHasUserRelationship(params: {
  sourceGroupEntity: Entity;
  targetUserEntity: Entity;
  groupMember: admin_directory_v1.Schema$Member;
}) {
  const { sourceGroupEntity, targetUserEntity, groupMember } = params;

  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: sourceGroupEntity,
    to: targetUserEntity,
    properties: {
      deliverySettings: groupMember.delivery_settings,
      email: groupMember.email,
      id: groupMember.id,
      kind: groupMember.kind,
      role: groupMember.role,
      status: groupMember.status,
      type: groupMember.type,
    },
  });
}

export function createAccountHasGroupRelationship(params: {
  accountEntity: Entity;
  groupEntity: Entity;
}) {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: params.accountEntity,
    to: params.groupEntity,
  });
}
