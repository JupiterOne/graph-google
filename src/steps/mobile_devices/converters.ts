import generateEntityKey from '../../utils/generateEntityKey';
import { admin_directory_v1 } from 'googleapis';
import { entities } from '../../constants';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
  Entity,
  RelationshipClass,
  createDirectRelationship,
} from '@jupiterone/integration-sdk-core';

function firstOrUndefined(v: any | any[]) {
  return Array.isArray(v) ? v[0] : undefined;
}

export function createMobileDeviceEntity(
  data: admin_directory_v1.Schema$MobileDevice,
) {
  const deviceName = data.deviceId as string;
  const name = firstOrUndefined(data.name);

  const serialNumber = data.serialNumber;
  const lastSeenOn = parseTimePropertyValue(data.lastSync);

  return createIntegrationEntity({
    entityData: {
      // The raw data on mobile devices can be very large, so we've removed it.
      source: {},
      assign: {
        _key: generateEntityKey(entities.MOBILE_DEVICE._type, deviceName),
        _type: entities.MOBILE_DEVICE._type,
        _class: entities.MOBILE_DEVICE._class,
        id: deviceName,
        name:
          name ||
          `${data.email?.[0] || 'Unknown User'}'s ${data.model || 'Device'}`,
        displayName:
          name ||
          `${data.email?.[0] || 'Unknown User'}'s ${data.model || 'Device'}`,
        category: 'mobile',
        make: data.model,
        model: data.model,
        serial: serialNumber,
        serialNumber,
        adbStatus: data.adbStatus,
        brand: data.brand,
        deviceCompromisedStatus: data.deviceCompromisedStatus,
        deviceId: data.deviceId,
        email: firstOrUndefined(data.email),
        encryptionStatus: data.encryptionStatus,
        firstSyncOn: parseTimePropertyValue(data.firstSync),
        hardware: data.hardware,
        hardwareId: data.hardwareId,
        imei: data.imei,
        lastSyncOn: lastSeenOn,
        lastSeenOn,
        manufacturer: data.manufacturer,
        ownerName: name,
        os: data.os,
        status: data.status,
        type: data.type,
        wifiMacAddress: data.wifiMacAddress,
        macAddress: data.wifiMacAddress,
        userAgent: data.userAgent,
      },
    },
  });
}

export function createAccountManagesMobileDeviceRelationship(params: {
  accountEntity: Entity;
  deviceEntity: Entity;
}) {
  return createDirectRelationship({
    _class: RelationshipClass.MANAGES,
    from: params.accountEntity,
    to: params.deviceEntity,
  });
}
