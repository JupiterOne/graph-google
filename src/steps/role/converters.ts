import generateEntityKey from '../../utils/generateEntityKey';
import { entities } from '../../constants';
import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';

// TODO/Rick: Interface is still for Account, update to role
interface CreateRoleEntityParams {
  account: { googleAccountId: string; name: string };
  domainNames: string[];
  primaryDomain?: string;
}

export function createRoleEntity(data: CreateRoleEntityParams) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: entities.ROLE._class,
        _key: generateEntityKey(
          entities.ROLE._type,
          data.account.googleAccountId,
        ),
        _type: entities.ROLE._type,
        // TODO/Rick: Shape
        displayName: data.account.name,
        name: data.account.name,
        domains: data.domainNames,
        primaryDomain: data.primaryDomain,
        accountId: data.primaryDomain,
        id: data.account.googleAccountId,
        vendor: 'Google',
      },
    },
  });
}
