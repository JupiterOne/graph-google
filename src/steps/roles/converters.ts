import generateEntityKey from '../../utils/generateEntityKey';
import { entities } from '../../constants';
import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { admin_directory_v1 } from 'googleapis';
import Schema$Role = admin_directory_v1.Schema$Role;

export function createRoleEntity(role: Schema$Role) {
  return createIntegrationEntity({
    entityData: {
      source: role,
      assign: {
        _class: entities.ROLE._class,
        _type: entities.ROLE._type,
        _key: generateEntityKey(entities.ROLE._type, role.roleId as string),
        // TODO/Rick: Is this the shape we want? I am including more properties for discussion but can whittle this down
        id: role.roleId as string,
        name: role.roleName,
        displayName: role.roleName as string,
        roleName: role.roleName,
        roleDescription: role.roleDescription,
        isSystemRole: role.isSystemRole,
        isSuperAdminRole: role.isSuperAdminRole,
        kind: role.kind,
        vendor: 'Google',
      },
    },
  });
}
