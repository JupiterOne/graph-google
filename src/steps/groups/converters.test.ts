import { admin_directory_v1, groupssettings_v1 } from 'googleapis';

import { getMockUser } from '../../../test/mocks';
import { createUserEntity } from '../users/converters';
import {
  createGroupEntity,
  createGroupHasGroupMappedRelationship,
  createGroupHasGroupRelationship,
  createGroupHasUserMappedRelationship,
  createGroupHasUserRelationship,
  createGroupSettingsEntity,
} from './converters';

function getMockGroup(
  partial?: admin_directory_v1.Schema$Group,
): admin_directory_v1.Schema$Group {
  return {
    kind: 'admin#directory#group',
    id: '0184mhaj0imamcg',
    etag: 'abc123',
    email: 'randomdistro@jupiterone.io',
    name: 'Random',
    directMembersCount: '5',
    description: 'randomdistro@jupiterone.io distribution list',
    adminCreated: true,
    aliases: ['randomdistro@jupiterone.com', 'randomdistro@jptr.one'],
    nonEditableAliases: ['randomdistro@jupiterone.io.test-google-a.com'],
    ...partial,
  };
}

function getMockGroupSettings(
  partial?: groupssettings_v1.Schema$Groups,
): groupssettings_v1.Schema$Groups {
  return {
    kind: 'groupsSettings#groups',
    email: 'homeschool@thewilliams.ws',
    name: 'homeschool',
    description: '',
    whoCanJoin: 'CAN_REQUEST_TO_JOIN',
    whoCanViewMembership: 'ALL_MANAGERS_CAN_VIEW',
    whoCanViewGroup: 'ALL_MEMBERS_CAN_VIEW',
    whoCanInvite: 'ALL_MANAGERS_CAN_INVITE',
    whoCanAdd: 'ALL_MANAGERS_CAN_ADD',
    allowExternalMembers: 'false',
    whoCanPostMessage: 'ANYONE_CAN_POST',
    allowWebPosting: 'true',
    maxMessageBytes: 26214400,
    isArchived: 'false',
    archiveOnly: 'false',
    messageModerationLevel: 'MODERATE_NONE',
    spamModerationLevel: 'MODERATE',
    replyTo: 'REPLY_TO_IGNORE',
    customReplyTo: '',
    includeCustomFooter: 'false',
    customFooterText: '',
    sendMessageDenyNotification: 'false',
    defaultMessageDenyNotificationText: '',
    showInGroupDirectory: 'false',
    allowGoogleCommunication: 'false',
    membersCanPostAsTheGroup: 'false',
    messageDisplayFont: 'DEFAULT_FONT',
    includeInGlobalAddressList: 'true',
    whoCanLeaveGroup: 'ALL_MEMBERS_CAN_LEAVE',
    whoCanContactOwner: 'ANYONE_CAN_CONTACT',
    whoCanAddReferences: 'NONE',
    whoCanAssignTopics: 'NONE',
    whoCanUnassignTopic: 'NONE',
    whoCanTakeTopics: 'NONE',
    whoCanMarkDuplicate: 'NONE',
    whoCanMarkNoResponseNeeded: 'NONE',
    whoCanMarkFavoriteReplyOnAnyTopic: 'NONE',
    whoCanMarkFavoriteReplyOnOwnTopic: 'NONE',
    whoCanUnmarkFavoriteReplyOnAnyTopic: 'NONE',
    whoCanEnterFreeFormTags: 'NONE',
    whoCanModifyTagsAndCategories: 'NONE',
    favoriteRepliesOnTop: 'true',
    whoCanApproveMembers: 'ALL_MANAGERS_CAN_APPROVE',
    whoCanBanUsers: 'OWNERS_AND_MANAGERS',
    whoCanModifyMembers: 'OWNERS_AND_MANAGERS',
    whoCanApproveMessages: 'OWNERS_AND_MANAGERS',
    whoCanDeleteAnyPost: 'OWNERS_AND_MANAGERS',
    whoCanDeleteTopics: 'OWNERS_AND_MANAGERS',
    whoCanLockTopics: 'OWNERS_AND_MANAGERS',
    whoCanMoveTopicsIn: 'OWNERS_AND_MANAGERS',
    whoCanMoveTopicsOut: 'OWNERS_AND_MANAGERS',
    whoCanPostAnnouncements: 'OWNERS_AND_MANAGERS',
    whoCanHideAbuse: 'NONE',
    whoCanMakeTopicsSticky: 'NONE',
    whoCanModerateMembers: 'OWNERS_AND_MANAGERS',
    whoCanModerateContent: 'OWNERS_AND_MANAGERS',
    whoCanAssistContent: 'NONE',
    customRolesEnabledForSettingsToBeMerged: 'false',
    enableCollaborativeInbox: 'false',
    whoCanDiscoverGroup: 'ALL_MEMBERS_CAN_DISCOVER',
    ...partial,
  };
}

function getMockGroupMemberTypeGroup(
  partial?: admin_directory_v1.Schema$Member,
): admin_directory_v1.Schema$Member {
  return {
    kind: 'admin#directory#member',
    etag: 'abc123',
    id: 'abc123',
    email: 'abc123email@jupiterone.com',
    role: 'MEMBER',
    type: 'GROUP',
    status: 'ACTIVE',
    ...partial,
  };
}

function getMockGroupMemberTypeUser(
  partial?: admin_directory_v1.Schema$Member,
): admin_directory_v1.Schema$Member {
  return {
    kind: 'admin#directory#member',
    etag: 'abc123',
    id: '1234567890',
    email: 'abc123email@jupiterone.com',
    role: 'MEMBER',
    type: 'USER',
    status: 'ACTIVE',
    ...partial,
  };
}

describe('#createGroupEntity', () => {
  test('should convert to entity', () => {
    expect(createGroupEntity(getMockGroup())).toMatchSnapshot();
  });
});

describe('#createGroupHasGroupMappedRelationship', () => {
  test('should convert to mapped relationship', () => {
    expect(
      createGroupHasGroupMappedRelationship(
        createGroupEntity(getMockGroup()),
        getMockGroupMemberTypeGroup(),
      ),
    ).toMatchSnapshot();
  });
});

describe('#createGroupHasUserMappedRelationship', () => {
  test('should convert to mapped relationship', () => {
    expect(
      createGroupHasUserMappedRelationship(
        createGroupEntity(getMockGroup()),
        getMockGroupMemberTypeUser(),
      ),
    ).toMatchSnapshot();
  });
});

describe('#createGroupHasGroupRelationship', () => {
  test('should convert to relationship', () => {
    expect(
      createGroupHasGroupRelationship({
        sourceGroupEntity: createGroupEntity(getMockGroup()),
        targetGroupEntity: createGroupEntity(getMockGroup({ id: '999' })),
        groupMember: getMockGroupMemberTypeGroup(),
      }),
    ).toMatchSnapshot();
  });
});

describe('#createGroupHasUserRelationship', () => {
  test('should convert to relationship', () => {
    expect(
      createGroupHasUserRelationship({
        sourceGroupEntity: createGroupEntity(getMockGroup()),
        targetUserEntity: createUserEntity(getMockUser()),
        groupMember: getMockGroupMemberTypeGroup(),
      }),
    ).toMatchSnapshot();
  });
});

describe('#createGroupSettingsEntity', () => {
  test('should convert to entity', () => {
    expect(
      createGroupSettingsEntity(getMockGroup(), getMockGroupSettings()),
    ).toMatchSnapshot();
  });
});
