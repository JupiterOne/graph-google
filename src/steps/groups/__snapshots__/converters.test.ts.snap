// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createGroupEntity should convert to entity 1`] = `
Object {
  "_class": Array [
    "UserGroup",
  ],
  "_key": "google_group_0184mhaj0imamcg",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "adminCreated": true,
        "aliases": Array [
          "randomdistro@jupiterone.com",
        ],
        "description": "randomdistro@jupiterone.io distribution list",
        "directMembersCount": "5",
        "email": "randomdistro@jupiterone.io",
        "etag": "abc123",
        "id": "0184mhaj0imamcg",
        "kind": "admin#directory#group",
        "name": "Random",
        "nonEditableAliases": Array [
          "randomdistro@jupiterone.io.test-google-a.com",
        ],
      },
    },
  ],
  "_type": "google_group",
  "adminCreated": true,
  "createdOn": undefined,
  "description": "randomdistro@jupiterone.io distribution list",
  "directMembersCount": "5",
  "displayName": "Random",
  "email": "randomdistro@jupiterone.io",
  "id": "0184mhaj0imamcg",
  "kind": "admin#directory#group",
  "name": "Random",
}
`;

exports[`#createGroupHasGroupMappedRelationship should convert to mapped relationship 1`] = `
Object {
  "_class": "HAS",
  "_key": "google_group_0184mhaj0imamcg_has_group_abc123",
  "_mapping": Object {
    "relationshipDirection": "FORWARD",
    "skipTargetCreation": false,
    "sourceEntityKey": "google_group_0184mhaj0imamcg",
    "targetEntity": Object {
      "_type": "google_group",
      "email": "abc123email@jupiterone.com",
    },
    "targetFilterKeys": Array [
      Array [
        "_type",
        "email",
      ],
    ],
  },
  "_type": "google_group_has_user",
  "displayName": "HAS",
}
`;

exports[`#createGroupHasGroupRelationship should convert to relationship 1`] = `
Object {
  "_class": "HAS",
  "_fromEntityKey": "google_group_0184mhaj0imamcg",
  "_key": "google_group_0184mhaj0imamcg|has|google_group_999",
  "_toEntityKey": "google_group_999",
  "_type": "google_group_has_group",
  "deliverySettings": undefined,
  "displayName": "HAS",
  "email": "abc123email@jupiterone.com",
  "id": "abc123",
  "kind": "admin#directory#member",
  "role": "MEMBER",
  "status": "ACTIVE",
  "type": "GROUP",
}
`;

exports[`#createGroupHasUserMappedRelationship should convert to mapped relationship 1`] = `
Object {
  "_class": "HAS",
  "_key": "google_group_0184mhaj0imamcg_has_user_1234567890",
  "_mapping": Object {
    "relationshipDirection": "FORWARD",
    "skipTargetCreation": false,
    "sourceEntityKey": "google_group_0184mhaj0imamcg",
    "targetEntity": Object {
      "_type": "google_user",
      "email": "abc123email@jupiterone.com",
    },
    "targetFilterKeys": Array [
      Array [
        "_type",
        "email",
      ],
    ],
  },
  "_type": "google_group_has_user",
  "displayName": "HAS",
}
`;

exports[`#createGroupHasUserRelationship should convert to relationship 1`] = `
Object {
  "_class": "HAS",
  "_fromEntityKey": "google_group_0184mhaj0imamcg",
  "_key": "google_group_0184mhaj0imamcg|has|google_user_123456789",
  "_toEntityKey": "google_user_123456789",
  "_type": "google_group_has_user",
  "deliverySettings": undefined,
  "displayName": "HAS",
  "email": "abc123email@jupiterone.com",
  "id": "abc123",
  "kind": "admin#directory#member",
  "role": "MEMBER",
  "status": "ACTIVE",
  "type": "GROUP",
}
`;

exports[`#createGroupHasUserRelationship should convert to relationship 2`] = `
Object {
  "_class": "HAS",
  "_fromEntityKey": "google_group_0184mhaj0imamcg",
  "_key": "google_group_0184mhaj0imamcg|has|google_user_123456789",
  "_toEntityKey": "google_user_123456789",
  "_type": "google_group_has_user",
  "deliverySettings": undefined,
  "displayName": "HAS",
  "email": "abc123email@jupiterone.com",
  "id": "abc123",
  "kind": "admin#directory#member",
  "role": "MEMBER",
  "status": "ACTIVE",
  "type": "GROUP",
}
`;
